src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILImporter.Interpreter.cs

```diff
@@ -239,22 +261,952 @@ private void ImportLoadFloat(double value)
+                            value = (int)((uint)value >> shiftBy);
+                        }
+
+                        _interpreter.EvaluationStack.Push(StackItem.FromInt32(value));
+                    }
+                    break;
+                case StackValueKind.Int64:
+                    {
+                        long value = op2.AsInt64();
+                        if (opcode == ILOpcode.shl)
+                        {
+                            value = value << shiftBy;
+                        }
+                        else if (opcode == ILOpcode.shr)
+                        {
+                            value = value >> shiftBy;
+                        }
+                        else if (opcode == ILOpcode.shr_un)
+                        {
+                            value = (long)((ulong)value >> shiftBy);
+                        }
+
+                        _interpreter.EvaluationStack.Push(StackItem.FromInt64(value));
+                    }
+                    break;
+                case StackValueKind.NativeInt:
+                    {
+                        IntPtr value = op2.AsNativeInt();
+                        if (opcode == ILOpcode.shl)
+                        {
+                            value = new IntPtr(value.ToInt64() << shiftBy);
+                        }
+                        else if (opcode == ILOpcode.shr)
+                        {
+                            value = new IntPtr(value.ToInt64() >> shiftBy);
+                        }
+                        else if (opcode == ILOpcode.shr_un)
+                        {
+                            UIntPtr uintPtr = (UIntPtr)value.ToPointer();
+                            value = new IntPtr((long)(uintPtr.ToUInt64() >> shiftBy));
+                        }
+
+                        _interpreter.EvaluationStack.Push(StackItem.FromNativeInt(value));
+                    }
+                    break;
+                case StackValueKind.Unknown:
+                case StackValueKind.Float:
+                case StackValueKind.ByRef:
+                case StackValueKind.ObjRef:
+                case StackValueKind.ValueType:
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
         }
 
         private void ImportCompareOperation(ILOpcode opcode)
         {
-            throw new NotImplementedException();
+            bool result = default(bool);
+            StackItem op1 = PopWithValidation();
+            StackItem op2 = PopWithValidation();
+
+            switch (op1.Kind)
+            {
+                case StackValueKind.Int32:
+                    {
+                        int val1 = op1.AsInt32();
+
+                        if (op2.Kind == StackValueKind.Int32 || op2.Kind == StackValueKind.NativeInt)
+                        {
+                            int val2 = op2.Kind == StackValueKind.Int32 ? op2.AsInt32() : op2.AsNativeInt().ToInt32();
```

Comment:
It is not correct to convert the NativeInt value to Int32 in this case. Instead, Int32 should be converted to Int64 before the comparison.

It may work better to first figure out the common type - this is what CppCodeGen does that:

```
// StackValueKind is carefully ordered to make this work (assuming the IL is valid)
StackValueKind kind = (op1.Kind > op2.Kind) ? op1.Kind : op2.Kind;
```

And then switch based on the common type.

Reply:
I addressed this by making `val2` a long and converting the NativeInt to that